### ISSUE TITLE
Changing how C target code accesses inputs, outputs, and actions

### ISSUE THREAD

**Issue by edwardalee** [icyphy/lingua-franca#184]
TLDR: The current syntax in reactions for accessing inputs, outputs, and actions are will make supporting multiports awkward. I would like to change it, but it won't be backward compatible. All existing C target code will require an update. Do we know how many users (if any) we have who will be affected? Specifically, in a reaction, to access input `in`, we would use the syntax `in->value` and `in->is_present` rather than `in` and `in_is_present`. For multiports, this will be `in[i]->value` and `in[i]->is_present`.

Any objections? Suggestions?

Why? Right now, there are several variants of `set()` macros in [reactor.h](https://github.com/icyphy/lingua-franca/blob/master/xtext/org.icyphy.linguafranca/src/lib/C/reactor.h). For example:
```
    output out:int;
    reaction(startup) {=
        set(out, 42);
    =}
```
The `set()` macro looks like this:
```
#define set(out, value) \
do { \
    out ## _is_present = true; \
    self->__ ## out = value; \
    self->__ ## out ## _is_present = true; \
} while(0)
```
This sets two elements of the self struct, `__out` and `__out_is_present` and one local variable `out_is_present` (the latter only for compatibility with inputs). The problem is that if the output is a multipart, I would like to say:
```
    output[4] out:int;
    reaction(startup) {=
        set(out[0], 42);
        set(out[1], 84);
        ...
    =}
```
There is no macro design (that I can see) that will handle this unless we change something fundamental. As usual, it means adding a level of indirection.  The proposed change is from a self struct with this:
```
    int out;
    bool out_is_present;
```
to a self struct with this:
```
    reactorClassName_out_t out;
```
where `reactorClassName_out_t` is given by a typedef:
```
    typedef struct {
        int value;
        bool is_present;
    } reactorClassName_out_t;
```
Instead of the local variable `out_is_present`, there would be a local variable `out` that is a pointer to this struct on the self struct.  The `set()` macro becomes simpler:
```
#define set(out, val) \
do { \
    out->value = val; \
    out->is_present = true; \
} while(0)
```
And supporting multiports becomes easier because there will just be an array of structs on the self struct.

**Comment by Soroosh129** [comment]
In `reactorClassName_out_t`, should the type of `value` be 'void *' or there is going to be a different struct for each type of value?

**Comment by Soroosh129** [comment]
An expansion with strong types could look like this:

```
struct reactorClassName_out_t {
    void * _value;
    bool is_present;
    void (*value)(const struct reactorClassName_out_t *);
}
```

And have value functions for each type:

```
int value_int(const struct reactorClassName_out_t* in)
{
    if(in->is_present)
        return (int)&_value; // Not sure how to correctly write this part
    else
        // Do something else
}
```

The construction would then look like

```
int value = 4;
struct reactorClassName_out_t port = { &value, false, value_int }; // Again not sure if &value will do the trick
printf("%d", port->value());
```

**Comment by Soroosh129** [comment]
Sorry my previous comment was incorrect and I have removed it. I had assumed you can use a function similar to print for strong types in C.

**Comment by lhstrh** [comment]
Yes, this makes sense to me. I'm not too worried about backward compatibility yet. This is a valid reason for a design change, and it only applies to the C target. Though it would be interesting to see what the impact of this change is on our (limited) set of benchmarks.

For convenience, we can provide shorthands for `in[i]->value` and `in[i]->is_present` using macros, e.g., `get(in[i])` and `check(in[i])`.

There is an alternative to consider. We don't have to use a struct; we can also use two arrays (I recall we used this design in our FMI implementation). In this case we'd have `in[i]` to retrieve the value and `in_is_present[i]` to check whether the value at index `i` is present. With this syntax, it would make sense to keep referring to `in` and `in_is_present` if `in` is not multiplex port. On a related note: how do we distinguish between singular ports that are of type array and multiplex ports?

**Comment by edwardalee** [comment]
Your alternative is what is currently implemented. The problem is that the set macros won't work with that alternative.  We will need a version of every one of these macros like the one I checked in:
```
     set_i(out, index, value);
```
In particular, there will be no way (that I see) to make this work:
```
    set(out[index], value);
```
As for array values vs. multiports, that's easy:
```
    int i = in[multiport_index]->value[array_index];
```

**Comment by lhstrh** [comment]
I see your point: I don't see a solution for the `set` macro either...

What do you think of adding the shorthand macros `get` and `check`? I think it looks a bit better...

**Comment by Soroosh129** [comment]
I would like to point out that it is possible to write the struct as:

    typedef struct {
        int _value;
        bool is_present;
        int (*get)(const struct reactorClassName_out_t *);
    } reactorClassName_out_t;

and have a function such as:

```
int value_int(const struct reactorClassName_out_t* in)
{
        if(in->is_present)
        {
                return in->_value;                                                                                                                                                                                                               }
        else
               // Do something else
}
```
You can basically initialize it and call this function like this:

```
int value = 4;
reactorClassName_out_t *port = malloc(sizeof(reactorClassName_out_t));
port->_value = value;
port->is_present = true;
port->get = value_int;


printf("%d\n", port->get(port));
```

I guess this is also an alternative but the caveat is that the struct instance needs to be passed to the get function.

**Comment by edwardalee** [comment]
Yes, I think these macros are good idea.
Not much of a shorthand, but pretty clear…

Edward

——————
Edward A. Lee
Professor of the Graduate School
EECS, UC Berkeley
http://eecs.berkeley.edu/~eal
